src/devices/sound/ymfm.cpp

```diff
@@ -0,0 +1,1663 @@
+	s8 lfo_raw_pm = clock_lfo();
+
+	// now update the state of all the channels and operators
+	for (int chnum = 0; chnum < RegisterType::CHANNELS; chnum++)
+		if (BIT(chanmask, chnum))
+			m_channel[chnum]->clock(m_env_counter, lfo_raw_pm, chnum == 2 && m_regs.multi_freq());
+
+	// return the envelope counter as it is used to clock ADPCM-A
+	return m_env_counter;
+}
+
+
+//-------------------------------------------------
+//  output - compute a sum over the relevant
+//  channels
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_engine_base<RegisterType>::output(s32 &lsum, s32 &rsum, u8 rshift, s16 clipmax, u8 chanmask) const
+{
+	// sum over all the desired channels
+	for (int chnum = 0; chnum < RegisterType::CHANNELS; chnum++)
+		if (BIT(chanmask, chnum))
+		{
+			// noise must be non-zero to use noise on OP4, so if it is enabled,
+			// OR with 2 (since only the LSB is actually checked for the noise state)
+			u8 noise = (chnum == 7 && m_regs.noise_enabled()) ? (m_noise_lfsr | 2) : 0;
+			m_channel[chnum]->output(m_lfo_am, noise, lsum, rsum, rshift, clipmax);
+		}
+}
+
+
+//-------------------------------------------------
+//  write - handle writes to the OPN registers
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_engine_base<RegisterType>::write(u16 regnum, u8 data)
+{
+	// most writes are passive, consumed only when needed
+	m_regs.write(regnum, data);
+
+	// handle writes to the mode register
+	if (regnum == RegisterType::REG_MODE)
+	{
+		// reset timer status
+		if (m_regs.reset_timer_b())
+			set_reset_status(0, STATUS_TIMERB);
+		if (m_regs.reset_timer_a())
+			set_reset_status(0, STATUS_TIMERA);
+
+		// load timers
+		update_timer(1, m_regs.load_timer_b());
+		update_timer(0, m_regs.load_timer_a());
+	}
+
+	// handle writes to the keyon registers
+	else if (regnum == RegisterType::REG_KEYON)
+	{
+		u8 chnum = m_regs.keyon_channel();
+		if (chnum < RegisterType::CHANNELS)
+			m_channel[chnum]->keyonoff(m_regs.keyon_states());
+	}
+}
+
+
+//-------------------------------------------------
+//  status - return the current state of the
+//  status flags
+//-------------------------------------------------
+
+template<class RegisterType>
+u8 ymfm_engine_base<RegisterType>::status() const
+{
+	u8 result = m_status & ~STATUS_BUSY;
+
+	// synthesize the busy flag if we're still busy
+	if (m_device.machine().time() < m_busy_end)
+		result |= STATUS_BUSY;
+	return result;
+}
+
+
+//-------------------------------------------------
+//  set_busy - set the busy flag in the status
+//  register
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_engine_base<RegisterType>::set_busy()
+{
+	// according to hardware tests, BUSY is set for 32 * prescaled clock
+	// regardless of actual data consumption
+	m_busy_end = m_device.machine().time() + attotime::from_hz(m_device.clock()) * (32 * m_clock_prescale);
```

Comment:
How often does it need to do this?  If it’s frequent, you’d be better off calculating the `attotime::from_hz(m_device.clock()) * (32 * m_clock_prescale)` part only on clock frequency changes.

Reply:
I pushed this out into the individual devices and precomputed as you suggest.